mknetid/mknetid.8
.\"     Title: mknetid
.\"    Author: 
.\" Generator: DocBook XSL Stylesheets v1.73.1 <http://docbook.sf.net/>
.\"      Date: 09/26/2007
.\"    Manual: NIS Reference Manual
.\"    Source: NIS Reference Manual
.\"
.TH "MKNETID" "8" "09/26/2007" "NIS Reference Manual" "NIS Reference Manual"
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.SH "NAME"
mknetid - generate data for netid map
.SH "SYNOPSIS"
.HP 20
\fB/usr/lib/yp/mknetid\fR [\-q] [\-h\ \fIhosts\fR] [\-p\ \fIpasswd\fR] [\-g\ \fIgroup\fR] [\-d\ \fIdomain\fR] [\-n\ \fInetid\fR]
.HP 20
\fB/usr/lib/yp/mknetid\fR \-\-version
.SH "DESCRIPTION"
.PP
\fBmknetid\fR
generates the
\fInetid\.byname\fR
NIS map from the contents of the group(5), passwd(5), hosts(5) and netid files\. It checks for multiple entrys of netids and warn for them or filters them out\. It is only called by
\fI/var/yp/Makefile\fR
when rebuilding the NIS map\.
.SH "OPTIONS"
.PP
\fB\-q\fR
.RS 4
This flag turns on \'quiet\' mode, don\'t print a warning message when finding an duplicate netid entry\.
.RE
.PP
\fB\-h\fR\fI hosts\fR
.RS 4
The
\fB\-h\fR
flag can be used to specify the use of another hosts file than the default
\fI/etc/hosts\fR\.
.RE
.PP
\fB\-p\fR\fI passwd\fR
.RS 4
The
\fB\-p\fR
flag can be used to specify the use of another passwd file than the default
\fI/etc/passwd\fR\.
.RE
.PP
\fB\-g\fR\fI group\fR
.RS 4
The
\fB\-g\fR
flag can be used to specify the use of another group file than the default
\fI/etc/group\fR\.
.RE
.PP
\fB\-n\fR\fI netid\fR
.RS 4
The
\fB\-n\fR
flag can be used to specify the use of another netid file than the default
\fI/etc/netid\fR\.
.RE
.PP
\fB\-d\fR\fI domain\fR
.RS 4
The
\fBmknetid\fR
command uses the system domainname by default\. If it is not set or you whish to override it, you must use the
\fB\-d\fR
parameter\.
.RE
.PP
\fB\-\-version\fR
.RS 4
Prints the version number
.RE
.SH "FILES"
.PP
\fI/etc/group\fR
.RS 4
groups file
.RE
.PP
\fI/etc/hosts\fR
.RS 4
hosts database
.RE
.PP
\fI/etc/netid\fR
.RS 4
netname database
.RE
.PP
\fI/etc/passwd\fR
.RS 4
password file
.RE
.SH "SEE ALSO"
.PP
\fBpasswd\fR(8),
\fBgroup\fR(5),
\fBhosts\fR(5),
\fBpasswd\fR(5)
.SH "AUTHOR"
.PP

\fBmknetid\fR
was written by Thorsten Kukuk <kukuk@linux\-nis\.org>\.


NEWS
ypserv NEWS -- history of user-visible changes.

Copyright (C) 1998-2015 Thorsten Kukuk

Please send bug reports, questions and suggestions to <kukuk@linux-nis.org>.

Version 3.7
* ypxfr: better error messages
* yppush: rewrote callback registration, enable --port again

Version 3.6
* yppush: register callback for IPv6, too.

Version 3.5
* yppush: convert more IPv4 only RPC functions to TI-RPC variants
* rpc.ypxfrd: Likewise.

Version 3.4
* yppush: convert IPv4 only RPC functions to TI-RPC variants

Version 3.3
* Adjust debug output for IPv6 addresses
* Fix yppush

Version 3.2.1
* ypserv.8 docu enhancements about securenets

Version 3.2
* Fix memory leak in access library

Version 3.1
* ypserv: Don't increase RPC buffer

Version 3.0
* Convert code to use ti-rpc functions
* Implement IPv6 support

Version 2.32.1
* Adjust for systemd >= 209

Version 2.32
* Update autoconf/automake
* ypserv: remove broken SIGHUP support.
* /var/yp/Makefile: read MINUID/MINGID from /etc/login.defs

Version 2.31
* Fix build failures

Version 2.30
* Add systemd support

Version 2.29
* Fix memory leaks in error case in rpc.yppasswdd and rpc.ypxfrd

Version 2.28
* Add support for tokyocabinet (gdbm replacement)
* Lot of small bug fixes

Version 2.27
* qdbm support was added
* Lot of small bug fixes

Version 2.26
* ypserv: Fix trusted_master option
* mknetid: Fix group parsing

Version 2.25
* Lot of small bug fixes/improved error handling

Version 2.24
* rpc.yppasswdd: Fix -x option

Version 2.23
* rpc.ypxfrd creates now a pid file

Version 2.22
* Update ypserv.8 manual page
* Fix possible memory corruption bug

Version 2.21
* Better handling of RPC decoding errors

Version 2.20
* Manual page cleanup
* Add some sanity checks

Version 2.19
* Fix ypxfr crashes
* Fix ypserv crash if own hostname is not resolveable

Version 2.18
* rpc.yppasswdd: Fix possible buffer overflow with -x/-E option

Version 2.17
* Fix debug output
* Don't ignore securenets for ypserv domain functions

Version 2.16
* Fix signal handlers
* Fix HP-UX support

Version 2.15
* makedbm: Use correct seperator if dumping databases
* rcp.yppasswdd, rpc.ypxfrd, yppush, ypserv: Compile with -fpie

Version 2.14
* ypxfr: Fix problems with fetching maps from SGI ypserv

Version 2.13
* ypserv: Set SLP domain attribute if configured
* ypserv: Fix memory leak if file cache is too small

Version 2.12.2
* ypserv: Refresh SLP registering after some time
* ypserv: Fixes for creating services.byservicename

Version 2.12.1
* ypserv: Change SLP URL to service:ypserv

Version 2.12
* ypserv: Fix compatibility problems with Solaris 9

Version 2.11
* ypserv: Add SLP support
* ypxfr: Fix seg.fault
* yppush: Add support for fix port

Version 2.10
* ypserv: Remove fork limit
* ypserv: Save/restore errno in signal handler
* makedbm: Remove spaces from end of key

Version 2.9
* ypxfr: query ypserv for correct master name
* Make it more portable
* Make number of allowed childs a compile time option
* Use system svc_run()
* Fix return code for wrong/invalid map- and domainname
* Make source code more portable

Version 2.8
* Fix compiler warnings
* Fix bug of counting running childs

Version 2.7
* ypserv now forks again for transfering complete NIS maps
* Update to latest autoconf/automake version

Version 2.6
* Fix bug that rpc.ypxfrd doesn't really close maps
* Allow user configureable number of cached database file handles

Version 2.5
* Fix memory leak

Version 2.4
* Minor bug fixes

Version 2.3
* Minor bug fixes

Version 2.2
* Minor bug fixes
* Allow IP numbers as master host
* Lookup master name of NIS map to avoid compare problems
* Fix problems with ypcat

Version 2.1
* Minor bug fixes/typos
* Fix paths in manual pages
* Document removal of tcp_wrapper support
* Fix bug so that ypxfr transfers the whole map in fallback mode
* rpc.yppasswdd can now register itself to a given port

Version 2.0
* Merge bug fixes from Debian

Version 1.99.0
* Rewrite

Version 1.3.10
* Fix various problems with SIGCHLD handling

Version 1.3.9
* Security update

Version 1.3.8
* yppush fix for ndbm support
* yppush fix for Irix 6.5.x NIS slave servers
* rpc.yppasswdd shadow rule fixes
* rpc.yppasswdd fixes for saving/restoring old file modes

Version 1.3.7
* All daemons give away controlling tty.
* Fix possible buffer overflow in rpc.yppasswdd.
* Use yp_msg and not syslog in rpc.yppasswdd for debug messages
* Use strcasecmp in yppush for comparing hostnames.
* Set Endianes for ypxfrd transfer correct.
* Fix file handle corruption through ypproc_all
* Fix error handling for fork() call
* Add support for keys and datum which are longer then 1024 bytes
* ypinit doesn't need a working ypwhich -m
* makedbm: some bug fixes for aliases and handling comments
* Add --debug flag to rpc.yppasswdd
* Add support for merging passwd with shadow file
* Add support for merging group with gshadow
* Minor bug fixes for 64bit platforms
* Bug fixes for libc5
* Add support for changing passwd file by external program

Version 1.3.6
* Fix bug in ypserv.conf parser
* Fix support for gdbm on 64bit platforms
* Fix yp_all return code for empty maps
* Upgrade ypmake to version 0.11 from Olaf Kirch
* Add support for passwd.adjunct
* Allow all characters >= 32

Version 1.3.5
* makedbm bug fixes

Version 1.3.4
* Fix output of ypxfr
* Fix /var/yp/Makefile dependency

Version 1.3.3
* Fix rpc.yppasswdd seg.fault in error case
* Make compiling with libc5 working again

Version 1.3.2
* Upate documentation
* Fix NDBM support

Version 1.3.1
* Bugfixes

Version 1.3.0
* Support for ndbm (SunOS 4.x/5.x) (Not complete)
* Add FQDN support (configure --enable-fqdn). All hostnames will
  go through gethostbyname to determine the offical hostname
* Add YP V1 protocol

Version 1.2.8
* Bugfixes

Version 1.2.7
* Bugfixes

Version 1.2.6
* pwupdate and makedbm man pages
* Remove tryresolve, hacker could stop ypserv with it

Version 1.2.5
* Add patches from SuSE distribution

Version 1.2.4
* yppasswd and rpc.yppasswdd bugfixes

Version 1.2.3

mknetid/mknetid.c
/* Copyright (c) 1996, 1999, 2001, 2002, 2009, 2012, 2014 Thorsten Kukuk
   Author: Thorsten Kukuk <kukuk@suse.de>

   The YP Server is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License
   version 2 as published by the Free Software Foundation.

   The YP Server is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   General Public License for more details.

   You should have received a copy of the GNU General Public
   License along with the YP Server; see the file COPYING. If
   not, write to the Free Software Foundation, Inc., 51 Franklin Street,
   Suite 500, Boston, MA 02110-1335, USA. */

/* mknetid - generate netid.byname map.  */

#if defined(HAVE_CONFIG_H)
#include "config.h"
#endif

#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <getopt.h>

#include "mknetid.h"

static int quiet_flag = 0;

char *
xstrtok (char *cp, int delim)
{
  static char *str = NULL;

  if (cp)
    str = cp;

  if (*str == '\0')
    return NULL;

  cp = str;

  if (delim == ' ')
    while (*str && (!isspace (*str)))
      str++;
  else
    while (*str && *str != delim)
      str++;

  if (*str)
    *str++ = '\0';

  return cp;
}

static void
Usage (int exitcode)
{
  fputs ("Usage: mknetid [-q] [-h hosts] [-p passwd] [-g group] [-d domain] [-n netid]\n",
	 stderr);
  fputs ("Usage: mknetid --version\n", stderr);
  exit (exitcode);
}

int
main (int argc, char *argv[])
{
  char *line = NULL;
  size_t length = 0;
  char *pwname = "/etc/passwd";
  char *grpname = "/etc/group";
  char *hostname = "/etc/hosts";
  char *netidname = "/etc/netid";
  char *domain = NULL;
  char domainbuf[1024];
  FILE *file;

  while (1)
    {
      int c;
      int option_index = 0;
      static struct option long_options[] =
      {
	{"version", no_argument, NULL, 'v'},
	{"host", required_argument, NULL, 'h'},
	{"group", required_argument, NULL, 'g'},
	{"passwd", required_argument, NULL, 'p'},
	{"netid", required_argument, NULL, 'n'},
	{"domain", required_argument, NULL, 'd'},
	{"quiet", no_argument, NULL, 'q'},
	{"help", no_argument, NULL, 'u'},
	{"usage", no_argument, NULL, 'u'},
	{NULL, 0, NULL, '\0'}
      };

      c = getopt_long (argc, argv, "uvqh:g:p:d:n:", long_options, &option_index);
      if (c == EOF)
	break;
      switch (c)
	{
	case 'q':
	  quiet_flag = 1;
	  break;
	case 'h':
	  hostname = optarg;
	  break;
	case 'g':
	  grpname = optarg;
	  break;
	case 'p':
	  pwname = optarg;
	  break;
	case 'n':
	  netidname = optarg;
	  break;
	case 'd':
	  domain = optarg;
	  break;
	case 'u':
	  Usage (0);
	  break;
	case 'v':
	  fprintf (stderr, "mknetid (%s) %s\n", PACKAGE, VERSION);
	  exit (0);
	default:
	  Usage (1);
	}
    }
  argc -= optind;
  argv += optind;

  if (argc != 0)
    Usage (1);

  if (domain == NULL)
    {
      if (getdomainname (domainbuf, sizeof (domainbuf)) != 0)
	{
	  fprintf (stderr, "YPPUSH: Cannot get default domain\n");
	  return 1;
	}
      else
	domain = domainbuf;
    }

  if ((file = fopen (pwname, "r")) == NULL)
    {
      fprintf (stderr, "ERROR: Can't open %s\n", pwname);
      exit (1);
    }

  while (!feof (file))
    {
      ssize_t n = getline (&line, &length, file);
      if (n < 1)
	break;

      if (line[0] != '+' && line[0] != '-')
	{

	  char *ptr, *key, *uid, *gid;
	  char *err_line = strdup (line);

	  key = xstrtok (line, ':');
	  ptr = xstrtok (NULL, ':');
	  uid = xstrtok (NULL, ':');
	  gid = xstrtok (NULL, ':');

	  if (key == NULL || ptr == NULL ||
	      uid == NULL || gid == NULL)
	    {
	      int n = strlen (err_line);

	      if (err_line[n - 1] == '\n')
		err_line[n - 1] = '\0';

	      if (strlen (err_line) > 0)
		fprintf (stderr, "WARNING: bad netid entry: '%s'\n",
			 err_line);
	      free (err_line);
	      continue;
	    }

	  free (err_line);

	  if (insert_user (key, domain, uid, gid) < 0)
	    if (!quiet_flag)
	      fprintf (stderr,
		       "WARNING: unix.%s@%s multiply defined, ignore new one\n",
		       uid, domain);
	}
    }
  fclose (file);

  if ((file = fopen (grpname, "r")) == NULL)
    {
      fprintf (stderr, "ERROR: Can't open %s\n", grpname);
      exit (1);
    }

  while (!feof (file))
    {
      ssize_t n = getline (&line, &length, file);
      if (n < 1)
	break;

      if (line[n - 1] == '\n')
        {
          if(n < 2)
            continue; /* empty line */
	  line[n - 1] = '\0';
        }

      if (line[0] != '+' && line[0] != '-')
	{
	  char *grpname, *ptr, *gid, *user;

	  grpname = xstrtok (line, ':');
	  ptr = xstrtok (NULL, ':');
	  gid = xstrtok (NULL, ':');
	  if(grpname && ptr && gid)
	    length -= gid - line;
	  else
	    {
	      fprintf (stderr, "ERROR: bad format of group \"%s\".\n",
		       grpname);
	      exit (1);
	    }
	  user = xstrtok (NULL, ',');
	  while (user != NULL) {
	    length = length - strlen(user) - 1;
	    if (add_group (user, gid) < 0)
	      if (!quiet_flag)
		fprintf (stderr, "WARNING: unknown user \"%s\" in group \"%s\".\n",
			 user, grpname);
	    if (length > 0)
	      user = xstrtok (user + strlen(user) + 1, ',');
	    else
	      user = NULL;
	  }
	}
    }
  fclose (file);

  if ((file = fopen (hostname, "r")) == NULL)
    {
      fprintf (stderr, "ERROR: Can't open %s\n", hostname);
      exit (1);
    }

  while (!feof (file))
    {
      ssize_t n = getline (&line, &length, file);
      if (n < 1)
	break;

      if (line[0] != '#')
	{
	  char *host;

	  (void)xstrtok (line, ' ');
	  host = xstrtok (NULL, ' ');
	  while (host != NULL && strlen (host) == 0)
	    host = xstrtok (NULL, ' ');

	  if (host != NULL)
	    if (insert_host (host, domain) < 0)
	      if (!quiet_flag)
		fprintf (stderr, "WARNING: unix.%s@%s multiply defined, ignore new one\n",
			 host, domain);
	}
    }

  fclose (file);

  print_table ();

  /*
  ** If /etc/netid does not exist, ignore it
  */
  if ((file = fopen (netidname, "r")) != NULL)
    {
      while (!feof (file))
	{
	  ssize_t n = getline (&line, &length, file);
	  if (n < 1)
	    break;

	  if (line[0] != '#')
	    {
	      int n = strlen (line);

	      if (line[n - 1] == '\n')
		line[n - 1] = '\0';
	      if (strpbrk (line, " \t") == NULL)
		{
		  if (strlen (line) > 0)
		    fprintf (stderr,
			     "WARNING: bad netid entry: '%s'\n",
			     line);
		}
	      else
		printf ("%s\n", line);
	    }
	}
      fclose (file);
    }

  free (line);
  return 0;
}
